[WJ-268] Merge ftml into monorepo

emmiegit opened the pull request:
This fulfills [FTML-27](https://scuttle.atlassian.net/browse/FTML-27), required by [WJ-268](https://scuttle.atlassian.net/browse/WJ-268), which aims to rewrite the Wikidot parser with the goals of maintainability and ease of future extension.

In order to preserve history and context around changes, as well as provide a blame, all the commits from the [original repo](https://github.com/Nu-SCPTheme/ftml) have been grafted into `develop` here. The commit history was rewritten so that they only affect the top-level directory `ftml`, which will contain files for this service. As such, this should be a merge or rebase commit, _not_ a squash commit. Also please avoid rebasing the PR, as this will interfere with imported version tags.

As described on its [README](https://github.com/Nu-SCPTheme/ftml/blob/master/README.md), `ftml` is a Rust library to parse and render Wikidot text instead of hacking changes into `Text_Wiki`. It also has a sub-crate, `ftml-http`, which provides a simple REST server to wrap the main calls the library provides. The crate does not use any unsafe code and has several tests on its abstract syntax tree (AST) generation to prevent regressions.

Its progress has been tracked on the [FTML](https://scuttle.atlassian.net/browse/FTML) board for some time.

emmiegit committed:
Add collapsible tests, and fix others.

emmiegit committed:
Add more collapsible AST tests.

emmiegit committed:
Address cargo clippy lints.

emmiegit committed:
Update dependencies.

emmiegit committed:
Create files for module parsing.

emmiegit committed:
Add ModuleRule and head method.

emmiegit committed:
Start parse_fn() function for ModuleRule.

emmiegit committed:
Implement ModuleRule boilerplate.

emmiegit committed:
Add module654 as an alias for module.

emmiegit committed:
Get module name using existing code.

emmiegit committed:
Invalid module test.

emmiegit committed:
Add module definition for enum Element.

emmiegit committed:
Add first module (backlinks).

emmiegit committed:
Rename module in rule.rs

emmiegit committed:
Add module categories.

emmiegit committed:
Add css module and null enum variant.

emmiegit committed:
Add rate module.

emmiegit committed:
Add join module.

emmiegit committed:
Add PageTree module.

emmiegit committed:
Export module definitions.

emmiegit committed:
Convert null modules into null elements.

emmiegit committed:
Replace Element::LineBreaks with tuple variant.

emmiegit committed:
Add name assertion helper functions.

emmiegit committed:
Add check_module_name().

emmiegit committed:
Rename functions to assert_*

emmiegit committed:
Pass in rules for convenience.

emmiegit committed:
Only check name assertions in debug mode.

emmiegit committed:
Replace block name assertions.

emmiegit committed:
Replace module name assertions.

emmiegit committed:
Add argument type retrieval helpers.

emmiegit committed:
Use new argument value methods.

emmiegit committed:
Add modules to mapping.

emmiegit committed:
Address warnings.

emmiegit committed:
Add test for very lage payloads (>200,000 chars).

emmiegit committed:
Update dependencies.

emmiegit committed:
Use kebab-case in module arguments.

emmiegit committed:
Fix collapsible tests.

emmiegit committed:
Change module serialization.

emmiegit committed:
Add kebab-case to child.

emmiegit committed:
Add initial module tests.

emmiegit committed:
Fix PageTree option parsing.

emmiegit committed:
Add variant module AST tests.

emmiegit committed:
Run cargo fmt.

emmiegit committed:
Shrink test payload for faster tests.

emmiegit committed:
Update dependencies.

emmiegit committed:
Add Element::Span variant.

emmiegit committed:
Add line break parsing for paragraph breaks in non-paragraph contexts.

emmiegit committed:
Add BLOCK_SPAN.

emmiegit committed:
Add BLOCK_SPAN to rule mapping.

emmiegit committed:
Fix main div block.

emmiegit committed:
Only remove leading and trailine newlines.

emmiegit committed:
Add span tests.

emmiegit committed:
Add elements for mark, ins, del.

emmiegit committed:
Add [[ins]] block.

emmiegit committed:
Add BLOCK_INS to mapping.

emmiegit committed:
Add BLOCK_DEL.

emmiegit committed:
Add BLOCK_DEL to block mapping.

emmiegit committed:
Add [[ins]] tests.

emmiegit committed:
Add [[del]] tests.

emmiegit committed:
Add alias tests.

emmiegit committed:
Add BLOCK_MARK.

emmiegit committed:
Add BLOCK_MARK to mapping.

emmiegit committed:
Add mark tests.

emmiegit committed:
Remove unnecessary null pruning.

emmiegit committed:
Add StyledContainer basic structures.

emmiegit committed:
Reorder definitions to be structure, then enum.

emmiegit committed:
Add documentation to each ContainerType variant.

emmiegit committed:
Add StyledContainerType variants.

emmiegit committed:
Move styled variants into StyledContainer.

emmiegit committed:
Fix StyledContainer invocations.

emmiegit committed:
Add html_tag() method.

emmiegit committed:
Add html_tag() method to HeadingLevel.

emmiegit committed:
Rename rules "bold" and "italics" to "strong" and "emphasis".

emmiegit committed:
Fix tests.

emmiegit committed:
Add html_attr() to AnchorTarget.

emmiegit committed:
Rename bold tests to strong.

emmiegit committed:
Rename italics tests to emphasis.

emmiegit committed:
Word.

emmiegit committed:
Rename ftml-server -> ftml-http.

emmiegit committed:
Add include module.

emmiegit committed:
Use page_ref in grammar instead.

emmiegit committed:
Start include implementation.

emmiegit committed:
Add start docs to README.

emmiegit committed:
Return included pages.

emmiegit committed:
Export include stuff in lib.rs

emmiegit committed:
Add INCLUDE_REGEX.

emmiegit committed:
Require matching full input in pest grammar.

emmiegit committed:
Start main include parsing code.

emmiegit committed:
Add parse() method to PageRef.

emmiegit committed:
Add docs to PageRef.

emmiegit committed:
Add docs to IncludeRef.

emmiegit committed:
Change PageRef construction.

emmiegit committed:
Fail parse on empty string.

emmiegit committed:
Trim input string in PageRef::parse.

emmiegit committed:
Add tests for PageRef::parse.

emmiegit committed:
Run cargo fmt.

emmiegit committed:
Add another test failure.

emmiegit committed:
Add todo stubs.

emmiegit committed:
Have Includer::include_pages() return Result.

emmiegit committed:
Add initial page text substitution.

emmiegit committed:
Progress towards parsing include blocks.

emmiegit committed:
Resolve fundamental borrowing issue.

emmiegit committed:
Actually append page for return.

emmiegit committed:
Change IncludeRef construction.

emmiegit committed:
Implement full argument parsing.

emmiegit committed:
Move includer to separate directory.

emmiegit committed:
Add DebugIncluder.

emmiegit committed:
Move include test to separate file.

emmiegit committed:
Uncomment tests.

emmiegit committed:
Fix DebugIncluder.

emmiegit committed:
Add missing import.

emmiegit committed:
Use DebugIncluder in tests.

emmiegit committed:
Add include test cases.

emmiegit committed:
Add uppercase include tests.

emmiegit committed:
Require one space after "include" in regex.

emmiegit committed:
Add debug logging for each regex match.

emmiegit committed:
Grab include content in regex lazily.

emmiegit committed:
Fix page ref order.

emmiegit committed:
Adjust logging message.

emmiegit committed:
Log on parse_include_block() error.

emmiegit committed:
Fix test case.

emmiegit committed:
Don't end pest on EOI always.

Preparing for start-based include parsing instead of range-based parsing.

emmiegit committed:
Fix tests.

emmiegit committed:
Resolve iterator clippy lint.

emmiegit committed:
Use Option<_> instead of Result<_, ()>.

emmiegit committed:
Require includes to be at the start of a line.

Beginning of start-based parsing instead of range-based parsing.

emmiegit committed:
Format trace!() call.

emmiegit committed:
Get parsed span instead of counting on regex ending.

emmiegit committed:
Run cargo fmt.

emmiegit committed:
Adjust comment.

emmiegit committed:
Add include substitution step to AST tests.

emmiegit committed:
Add BLOCK_INCLUDE.

emmiegit committed:
Add block to mapping.

emmiegit committed:
Run cargo fmt.

emmiegit committed:
Add page name to DebugIncluder.

emmiegit committed:
Add first include AST test.

emmiegit committed:
Add trace log for post-get_block_name().

emmiegit committed:
Fix bug with special block name fetching.

emmiegit committed:
Run cargo fmt.

emmiegit committed:
Get block name in invalid special warnings.

emmiegit committed:
Add base include AST tests.

emmiegit committed:
Add first include parameters AST test.

emmiegit committed:
Add MapWrap for DebugIncluder.

emmiegit committed:
Use single quotes in key-value output.

emmiegit committed:
Add tests for MapWrap.

emmiegit committed:
Use formatted MapWrap in output.

emmiegit committed:
Use regular rust escaping for strings.

Limit text elements in AST JSON files.

emmiegit committed:
Fix AST tests for new map formatting.

emmiegit committed:
Add tests for parameter separators.

emmiegit committed:
Add newline and misc include AST tests.

emmiegit committed:
Add multiline include block AST test.

emmiegit committed:
Fix map_wrap tests.

emmiegit committed:
Fix AST tests with MapWrap formatting.

emmiegit committed:
Update ftml::include() signature in README.

emmiegit committed:
Fix old ftml-server references.

emmiegit committed:
Add include route to ftml-http.

emmiegit committed:
Add include to router.

emmiegit committed:
Set max content length to 4 MiB.

emmiegit committed:
Add TODO for adding include.

emmiegit committed:
Start HttpIncluder implementation.

emmiegit committed:
Stub Includer trait impl.

emmiegit committed:
Return Result in no_such_include().

emmiegit committed:
Handle error in no_such_include().

emmiegit committed:
Implement initial include route processing.

emmiegit committed:
Stubbed request to callback URL.

emmiegit committed:
Request pages from callback.

emmiegit committed:
Add openssl library in build, if relevant.

emmiegit committed:
Use list of fetched pages instead of un-JSON map.

emmiegit committed:
Add ftml_includer_server.py to allow testing ftml-http.

emmiegit committed:
Log reqwest error properly.

emmiegit committed:
Resolve compiler warning.

emmiegit committed:
Change field name in IncludeRef.

emmiegit committed:
Fix python includer http server script.

emmiegit committed:
Fix PageRef rendering.

emmiegit committed:
Add message to ftml_includer_server.py

emmiegit committed:
Use wrapper structure for proper outputs.

emmiegit committed:
Split objects into submodule.

emmiegit committed:
Move include route to separate submodule.

Preparing for other routes to move.

emmiegit committed:
Move preprocess route to separate submodule.

emmiegit committed:
Move tokenize route to separate submodule.

emmiegit committed:
Move parse route to separate submodule.

emmiegit committed:
Move render route to separate submodule.

emmiegit committed:
Resolve compiler warnings.

emmiegit committed:
Move miscellaneous route to separate submodule.

emmiegit committed:
Add newline.

emmiegit committed:
Move process_include() out.

emmiegit committed:
Rename process_include -> run_include.

emmiegit committed:
Adapt run_include().

emmiegit committed:
Run include substitution before preprocess.

emmiegit committed:
Begin extracting results to manage lifetimes.

emmiegit committed:
Collapse route_tokenize().

emmiegit committed:
Fix compilation.

emmiegit committed:
Start route_parse() with includes.

emmiegit committed:
Borrow tokens instead of passing ownership.

emmiegit committed:
Implement route_render_html().

emmiegit committed:
Finish route_render_html().

emmiegit committed:
Resolve clippy lint.

emmiegit committed:
Change ftml_includer_server.py logging.

emmiegit committed:
Make some more modules public for exposing organized items.

emmiegit committed:
Move modules to allow for top-level exports.

emmiegit committed:
Fix ftml-http build.

emmiegit committed:
Fix tests.

emmiegit committed:
Remove unused TextInput.

emmiegit committed:
Update documentation.

emmiegit committed:
Expose HtmlRenderOutput.

emmiegit committed:
Have all ftml-http outputs be kebab-case.

emmiegit committed:
Add route_render_debug().

emmiegit committed:
Add to route.

emmiegit committed:
Fix ftml-http typing.

emmiegit committed:
Remove defunct include blocked tests.

emmiegit committed:
Rename parser -> parsing, includer -> includes.

Resolve clippy lint.

emmiegit committed:
Fix build for ftml-http.

emmiegit committed:
Fix tests.

emmiegit committed:
Zip fetched pages with requests.

emmiegit committed:
Construct invalid response error.

emmiegit committed:
Always return fetched page in server.

emmiegit committed:
Use page-ref in sample ftml includer server.

emmiegit committed:
PageRef serializes itself as kebab-case.

emmiegit committed:
Have IncludeRef use kebab-case too.

emmiegit committed:
Have FetchedPage use kebab-case too...

emmiegit committed:
Fix tests.

emmiegit committed:
Run cargo fmt.

emmiegit committed:
Update docs with include entry order change.

emmiegit committed:
Update dependencies.

emmiegit committed:
Bump version to v0.4.0

emmiegit committed:
Update rustc version in README.

emmiegit committed:
Replace copyright header to say "Wikijump Team".

emmiegit committed:
Add new URL schemes.

emmiegit committed:
Add {$name} variables to lexer.

emmiegit committed:
Add AST test for variable.

As well as string, which didn't have one.

emmiegit committed:
Add variable to lexer rule.

lol

emmiegit committed:
Use atomic rules in lexer.

Should be slightly faster because it doesn't have to generate internal
pairs.

rossjrw commented on ftml/.github/workflows/ci.yaml:
```diff
@@ -0,0 +1,64 @@
+on:
```
I _think_ any Actions config needs to be in the top-level .github directory

emmiegit replied:
Yeah, this just copies all the commits from the repo, there are some things to do afterwards with adjusting CI and getting it integrated into Docker.

emmiegit commented:
https://scuttle.atlassian.net/browse/FTML-27

https://scuttle.atlassian.net/browse/WJ-268

rossjrw reviewed (approved):
It's good to see this chunk of the project joining the rest of its family.

emmiegit committed:
Move CI workflow file.

emmiegit committed:
Consistent naming for CI steps.

Scheme is "[Service] Description"
For services wthout specific names, their programming language is
referenced instead.

emmiegit committed:
Use *.yaml

emmiegit committed:
Shorten CI job names.

emmiegit committed:
Use direct execution for ftml CI.

Allows changing directory first.

emmiegit committed:
Re-include toolchain.

emmiegit commented:
Fixed the CI, whenever the service is modified the builds will fire.

emmiegit committed:
Move documentation to new top-level directory.

emmiegit committed:
Merge ftml gitignore with root.

emmiegit committed:
Update CI badge.

emmiegit committed:
Tweaks to README.

emmiegit committed:
Change cargo specification.

emmiegit committed:
Remove line about crate uploads.

emmiegit committed:
Fix JSON examples.

Commas...

emmiegit commented:
Documentation and such should now be updated.

rossjrw commented on ftml/docs/README.md:
```diff
@@ -0,0 +1 @@
+../README.md
```
Curious - what's this link here for?

emmiegit replied:
I guess since it's documentation. It's not really necessary, I just thought I'd add it.

rossjrw commented on .github/workflows/test-js.yaml:
```diff
@@ -1,4 +1,5 @@
-name: '[Javascript] tests'
+name: '[js] Tests'
```
Given that the FTML badge title is `[<project>] language`, I'd prefer something like `]frontend] JS` or `[web] JS` for consistency. The URL for the badge on the top README will also need to be updated

rossjrw commented on ftml/docs/Philosophy.md:
```diff
@@ -0,0 +1,36 @@
+edge-case conditions.
+
+This rewrite of ftml performs preprocessing substitions and tokenization like the first version,
+but the parser is hand-written to allow for loose fallback rules.
```
I'm not sure the distinction between the current version of ftml and the original version, while historically important in terms of code history and developer experience, important reading for what I expect to be the average reader of this document. I feel like the philosophy of fallbacks could be described with the assumption that the current version is the only canonical version

emmiegit committed:
Update labels for consistency.

emmiegit committed:
Update top-level README badges.

emmiegit committed:
Update philosophy documentation, this is only canonical ftml.

emmiegit committed:
Run test-js whenever the workflow file is updated.

emmiegit committed:
Adjust JS test workflow name.

emmiegit commented:
thanks @rossjrw
